### PR TITLE
3423 include with client assigned id partition

### LINKED ISSUE #3423: _include not working with client-assigned ID in partitioned environment
NOTE: Before filing a ticket, please see the following URL:
https://github.com/hapifhir/hapi-fhir/wiki/Getting-Help

**Describe the bug**
Have two partitions called ‘part1' and ‘part2’. The plan is to load the Patient resource in DEFAULT partition with client assigned resource id based on member id, and load the ADT data (Encounter, Practitioner, Care Team and Patient will be excluded) in ‘part1’ partition with the Patient reference with client assigned id (which will exists in DEFAULT partition as part of pre-load) and Condition (risk scores) resource as in 'part2’ partition.
If we want to get only Condition ('part2' partition) that include Patient, we are not able to read through part2 endpoint since Patient does not exist in ‘part2’ partition. Same use-case applies for exclusive 'part1' partition.

**To Reproduce**
1. Create a partition (e.g. PARTITION-123).
2. PUT a Patient to the DEFAULT partition, using a custom ID (e.g. "id": "test"). 
PUT http://localhost:8000/DEFAULT/Patient/test
3. POST a Condition to the PARTITION-123 partition, with a reference to the Patient created above. 
POST http://localhost:8000/PARTITION-123/Condition 
4. Perform this GET:
http://localhost:8000/PARTITION-123/Condition?subject=Patient/test&_include=Condition:subject

**Expected behavior**
The _include query works using both server-assigned IDs and client-assigned IDs

**Screenshots**

**Environment (please complete the following information):**
 - HAPI FHIR Version

**Additional context**

### PR DESCRIPTION
**Pull request by samuelwlee2**
When the Cross-Partition Reference Mode is set to ALLOWED_UNQUALIFIED  in the partition environment, it is searching for the wrong partition id and resulted in resource persistent id not found. Fixed the code by searching across all the partitions when ALLOWED_UNQUALIFIED is set.

closes #3423

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java [hunk 1]
```diff
@@ -341,4 +344,31 @@ public void testDirectDaoAccess_PartitionInRequestDetails_TransactionWithGet() {
...
 
 	}
+
+	@Test
+	public void testIncludeInTenantWithAssignedID() throws Exception {
+		IIdType idA = createResource("Patient", withTenant(JpaConstants.DEFAULT_PARTITION_NAME), withId("test"), withFamily("Smith"), withActiveTrue());
+		createConditionWithAllowedUnqualified(idA);
+		Bundle response = myClient.search().byUrl(myClient.getServerBase() + "/" + TENANT_A + "/Condition?subject=Patient/" + idA.getIdPart() + "&_include=Condition:subject").returnBundle(Bundle.class).execute();
+		assertThat(response.getEntry().size(), is(equalTo(2)));
+		myPartitionSettings.setAllowReferencesAcrossPartitions(PartitionSettings.CrossPartitionReferenceMode.NOT_ALLOWED);
```
Comment (oliviayou) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java:354]: Nice test!  I'm wondering why you set CrossPartitionReferenceMode to `NOT_ALLOWED` at the end of the test?

Comment (samuelwlee2) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java:354]: This partition setting is used by other tests, and I do not want this flag to effect other tests as I don't believe this setting is reseted back after the test.

Comment (oliviayou) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java:354]: Makes sense 👍🏻

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java [hunk 2]
```diff
@@ -311,23 +312,7 @@ private void doResolvePersistentIds(RequestPartitionId theRequestPartitionId, Li
...
-				if (partitionIds.size() > 1) {
-					Predicate partitionIdCriteria = from.get("myPartitionIdValue").as(Integer.class).in(partitionIds);
-					andPredicates.add(partitionIdCriteria);
-				} else {
-					Predicate partitionIdCriteria = cb.equal(from.get("myPartitionIdValue").as(Integer.class), partitionIds.get(0));
-					andPredicates.add(partitionIdCriteria);
-				}
-			}
-
+			getOptionalPartitionPredicate(theRequestPartitionId, cb, from).ifPresent(andPredicates::add);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java:315]: thank you for pulling this out to a separate method!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java [hunk 3]
```diff
@@ -346,9 +331,33 @@ private void doResolvePersistentIds(RequestPartitionId theRequestPartitionId, Li
...
 	}
 
+	/**
+	 * Return optional predicate for searching on forcedId
+	 * 1. If the partition mode is ALLOWED_UNQUALIFIED, the return optional predicate will be empty, so search is across all partitions.
+	 * 2. If it is default partition and default partition id is null, then return predicate for null partition.
+	 * 3. If the requested partition search is not all partition, return the request partition as predicate.
+	 */
+	private Optional<Predicate> getOptionalPartitionPredicate(RequestPartitionId theRequestPartitionId, CriteriaBuilder cb, Root<ForcedId> from) {
+		if (myPartitionSettings.getAllowReferencesAcrossPartitions().equals(PartitionSettings.CrossPartitionReferenceMode.ALLOWED_UNQUALIFIED)) {
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java:343]: Please move this into a helper method inside PartitionSettings.  I.e.
`if (myPartitionSettings.isAllowUnqualifiedCrossPartitionReference())` or something like that

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java [hunk 4]
```diff
@@ -346,9 +331,33 @@ private void doResolvePersistentIds(RequestPartitionId theRequestPartitionId, Li
...
+			Predicate partitionIdCriteria = cb.isNull(from.get("myPartitionIdValue").as(Integer.class));
+			return Optional.of(partitionIdCriteria);
+		} else if (!theRequestPartitionId.isAllPartitions()) {
+			List<Integer> partitionIds = theRequestPartitionId.getPartitionIds();
+			partitionIds = replaceDefaultPartitionIdIfNonNull(myPartitionSettings, partitionIds);
+			if (partitionIds.size() > 1) {
+				Predicate partitionIdCriteria = from.get("myPartitionIdValue").as(Integer.class).in(partitionIds);
+				return Optional.of(partitionIdCriteria);
+			} else {
+				Predicate partitionIdCriteria = cb.equal(from.get("myPartitionIdValue").as(Integer.class), partitionIds.get(0));
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/index/IdHelperService.java:355]: What if `partitionIds` is empty?

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java [hunk 5]
```diff
@@ -341,4 +344,31 @@ public void testDirectDaoAccess_PartitionInRequestDetails_TransactionWithGet() {
...
+	}
+
+	@Test
+	public void testIncludeInTenantWithAutoGeneratedID() throws Exception {
+		// Create patient
+		IIdType idA = createResource("Patient", withTenant(JpaConstants.DEFAULT_PARTITION_NAME), withFamily("Smith"), withActiveTrue());
+		createConditionWithAllowedUnqualified(idA);
+		Bundle response = myClient.search().byUrl(myClient.getServerBase() + "/" + TENANT_A + "/Condition?subject=Patient/" + idA.getIdPart() + "&_include=Condition:subject").returnBundle(Bundle.class).execute();
+		assertThat(response.getEntry().size(), is(equalTo(2)));
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java:363]:
```suggestion
		assertThat(response.getEntry(), hasSize(2));
```

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java [hunk 6]
```diff
@@ -341,4 +344,31 @@ public void testDirectDaoAccess_PartitionInRequestDetails_TransactionWithGet() {
...
+		myPartitionSettings.setAllowReferencesAcrossPartitions(PartitionSettings.CrossPartitionReferenceMode.NOT_ALLOWED);
```
Comment (fil512) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/provider/r4/MultitenantServerR4Test.java:364]: this should either be in a finally block, or better yet in an @AfterEach method

**Review by fil512 — APPROVED** [review]

**Pull request merged**